### ISSUE TITLE
Grooming Sprint #35

### ISSUE THREAD

**Issue by vttranlina** [linagora/james-project#4912]
- Postgres Message Mapper ...
- ACLs (array in mailbox table?)
- ...

**Comment by Arsnael** [comment]
The tasks I could perhaps see:
* AbstractJPAMailboxMessage -> AbstractPostgresMailboxMessage POJO (+PostgresProperty and PostgresUserFlag)
* JPAEncryptedMailboxMessage, JPAMailboxMessage, JPAMailboxMessageWithAttachmentStorage (+PostgresAttachment) POJOs
(* PostgresAttachmentMapper ?)
(* JPAStreamingMailboxMessage -> PostgresStreamingMailboxMessage ?)
* PostgresUidProvider and PostgresModSeqProvider
* PostgresMessageMapper
* MailboxACL -> array in PostgresMailbox
* Guice bindings for PosgresMessageMapper
* Perf tests

The ones between () are tasks that I'm not sure are necessary or not

**Comment by vttranlina** [comment]
<details>
<summary>Sample code with Array data type</summary>

```java

 @Test
    void arrayTest() {
        Table<Record> tableName = DSL.table("tbn1");
        Field<String> clm1 = DSL.field("clm1", SQLDataType.VARCHAR(255).notNull());
        Field<String[]> clm4 = DSL.field("clm4", SQLDataType.VARCHAR.getArrayDataType());

        PostgresTable table = PostgresTable.name(tableName.getName())
            .createTableStep((dsl, tbn) -> dsl.createTable(tbn)
                .column(clm1)
                .column(clm4)
            )
            .noRLS();

        PostgresModule module = PostgresModule.table(table);
        PostgresTableManager testee = tableManagerFactory.apply(module);

        testee.initializeTables()
            .block();

        // insert data {a,b,c} -> db = {a,b,c}
        postgresExtension.getPostgresExecutor()
            .executeVoid(dsl -> Mono.from(dsl.insertInto(tableName, clm1, clm4)
                .values("tung", new String[]{"a", "b", "c"})
                .returningResult(clm1, clm4)))
            .block();

        // update data {d,e,f} -> db = {d,e,f}
        postgresExtension.getPostgresExecutor()
            .executeVoid(dsl -> Mono.from(dsl.update(tableName)
                .set(clm4, new String[]{"d", "e", "f"})
                .where(clm1.eq("tung"))
                .returningResult(clm1, clm4)))
            .block();

        // append `g` to the array -> db = {d,e,f,g}
        postgresExtension.getPostgresExecutor()
            .executeVoid(dsl -> Mono.from(dsl.update(tableName)
                .set(clm4, PostgresDSL.arrayAppend(clm4, "g"))
                .where(clm1.eq("tung"))
                .returningResult(clm1, clm4)))
            .block();

        // remove `d` from the array -> db = {e,f,g}
        postgresExtension.getPostgresExecutor()
            .executeVoid(dsl -> Mono.from(dsl.update(tableName)
                .set(clm4, PostgresDSL.arrayRemove(clm4, "d"))
                .where(clm1.eq("tung"))
                .returningResult(clm1, clm4)))
            .block();

        // append array (`x`, `y`) to the array -> db = {e,f,g,x,y}
        postgresExtension.getPostgresExecutor()
            .executeVoid(dsl -> Mono.from(dsl.update(tableName)
                .set(clm4, PostgresDSL.arrayCat(clm4, new String[]{"x", "y"}))
                .where(clm1.eq("tung"))
                .returningResult(clm1, clm4)))
            .block();
    }
```
</details>

Ref: array function https://www.postgresql.org/docs/current/functions-array.html

**Comment by chibenwa** [comment]
### Remark 1

`AbstractPostgresMailboxMessage` -> :red_circle: 

`PAEncryptedMailboxMessage, JPAMailboxMessage, JPAMailboxMessageWithAttachmentStorage (+PostgresAttachment) POJOs` -> :red_circle: 

Just use reuse the existing data model and do not create something custom.

Use `new MessageFactory.StoreMessageFactory()`.

Also forget emcrypted / streaming capabilities from JPA: it is big crap.

So `(JPAStreamingMailboxMessage -> PostgresStreamingMailboxMessage ?)`  -> :red_circle: 

### Remark 2 

`(PostgresAttachmentMapper ?)` -> Not needed for IMAP

### Remark 3

`MailboxACL -> array in PostgresMailbox` -> Not already handled in mailbox ticket of current sprint?

### Misc

Most work will be implementing message mapper. Maybe we can consider implementing things like `UsersRepository`, `DomainRepository`, `SieveRepository` and `RecipientRewriteTable` in order to fill the gaps?

**Comment by vttranlina** [comment]
> MailboxACL -> array in PostgresMailbox -> Not already handled in mailbox ticket of current sprint?

yes

**Comment by vttranlina** [comment]
> PostgresUidProvider and PostgresModSeqProvider

Should we use `Sequence` for it? https://www.postgresql.org/docs/current/sql-createsequence.html

+ we don't need a dedicate table for UidProvider
+ we don't need query get last uid, and ++ for nextUid, and query update newUid to database
- The sequence will be shared for all mailbox. -> Old: MailboxA: 1,2,3.... MailboxB: 1,2,3....   New: MailboxA: 1,3,4... MailboxB: 2,5,6...

**Comment by Arsnael** [comment]
> https://github.com/linagora/james-project/issues/4897#issuecomment-1780535240

We agreed on an other ticket, and this sprint was already full with the tickets we have. In my head it was for the next sprint. But maybe we could reorganize?

**Comment by quantranhong1999** [comment]
> Most work will be implementing message mapper. Maybe we can consider implementing things like UsersRepository, DomainRepository, SieveRepository and RecipientRewriteTable in order to fill the gaps?

#### PostgresDomainDAO + PostgresDomainList (should be implement before UsersRepository because UsersRepository need DomainList)
    -> Create a domain table:
    CREATE TABLE IF NOT EXISTS public.james_domain
    (
        domain_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
        CONSTRAINT james_domain_pkey PRIMARY KEY (domain_name)
    )

    Dod: pass contract test + guice binding

#### PostgresUsersDAO + PostgresUsersRepository
    -> Create a users table:
    CREATE TABLE IF NOT EXISTS public.james_user
    (
        user_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
        hashed_password character varying(128) COLLATE pg_catalog."default" NOT NULL,
        password_hash_algorithm character varying(100) COLLATE pg_catalog."default" NOT NULL,
        CONSTRAINT james_user_pkey PRIMARY KEY (user_name)
    )

    AUTHORIZED_USERS + DELEGATED_USERS -> could be done later upon DelegationStore ticket.

    Dod: pass contract test + guice binding

#### PostgresSieveQuotaDAO + PostgresSieveScriptDAO + PostgresSieveRepository
    -> Create a SieveQuota table:
    CREATE TABLE IF NOT EXISTS public.james_sieve_quota
    (
        user_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
        size bigint NOT NULL,
        CONSTRAINT james_sieve_quota_pkey PRIMARY KEY (user_name)
    )

    -> Create a SieveScript table:
    CREATE TABLE IF NOT EXISTS public.james_sieve_script
    (
        activation_date_time timestamp with time zone,
        is_active boolean NOT NULL,
        script_content character varying(1024) COLLATE pg_catalog."default" NOT NULL,
        script_name character varying(255) COLLATE pg_catalog."default" NOT NULL,
        script_size bigint NOT NULL,
        user_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
        CONSTRAINT james_sieve_script_pkey PRIMARY KEY (user_name, script_name)
    )

    DoD: pass contract tests + guice binding

#### PostgresRecipientRewriteTableDAO + PostgresRecipientRewriteTable
    -> Create a rrt table:
    CREATE TABLE IF NOT EXISTS public.james_recipient_rewrite
    (
        domain_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
        user_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
        target_address character varying(100) COLLATE pg_catalog."default" NOT NULL,
        CONSTRAINT james_recipient_rewrite_pkey PRIMARY KEY (domain_name, user_name)
    )

    DoD: pass contract tests + guice binding

**Comment by chibenwa** [comment]
>> PostgresUidProvider and PostgresModSeqProvider
> Should we use Sequence for it? https://www.postgresql.org/docs/current/sql-createsequence.html

:red_circle: doing a sequence for all mailboxes imposes linearizability constraints that each of our updates would (needlessly) need to satisfy, thus pausing a needless scalability / performance constraint. By design we cannot afford have a bottleneck centralizing all our write operations. I thus strongly disagree using *sequences*.

**Comment by chibenwa** [comment]
> PostgresDomainDAO + PostgresDomainList (should be implement before UsersRepository because UsersRepository need DomainList)

Username length is up to 255 chars...

> AUTHORIZED_USERS + DELEGATED_USERS -> could be done later upon DelegationStore ticket.

Ticket for it too?

> PostgresSieveQuotaDAO

Please to like Cassandra: 1 DAO for quota?

> CREATE TABLE IF NOT EXISTS public.james_sieve_script

script_content character varying(1024) COLLATE pg_catalog."default" NOT NULL,

=> TEXT. We do not want to limit ourselves at 1KB...

`activation_date_time timestamp with time zone,`

=> Store dates in UTC?

username is up to 255 chars...

> RRT

```
    domain_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
    user_name character varying(100) COLLATE pg_catalog."default" NOT NULL,
```

Domain names are up to 253 chars. Usernames are up to 255 chars.

**Comment by vttranlina** [comment]
@chibenwa thinking again about mailbox ACLs, is an array enough?

Mailbox ACLs have also entry keys (user, group, special) so did you mean having a very naive implementation perhaps (like only 1 entry key by default so we just store an array of rights) or do we need maybe a map here? (we ain't sure if we can have a map<String, Array> in postgres)

**Comment by vttranlina** [comment]
@chibenwa (Rene) I thought sprints were supposed to be half postgres and not full, should we really treat all those mappers? We could though still groom it and sort it out during planning

**Comment by vttranlina** [comment]
> Most work will be implementing message mapper. Maybe we can consider implementing things like UsersRepository, DomainRepository, SieveRepository and RecipientRewriteTable in order to fill the gaps?

(Rene again) shouldn't we try to migrate fully pg-mailbox from jpa to postgres first before starting migrating other mappers in other modules?

**Comment by chibenwa** [comment]
> @chibenwa thinking again about mailbox ACLs, is an array enough?

hstore could do too

> @chibenwa (Rene) I thought sprints were supposed to be half postgres and not full, should we really treat all those mappers? We could though still groom it and sort it out during planning

I do not mind to burn a full sprint to get a 100% PG James IMAP server as a christmas gift ;-)

**Comment by Arsnael** [comment]
> I do not mind to burn a full sprint to get a 100% PG James IMAP server as a christmas gift ;-)

Do I need to buy Santa's suit, let my beard grow and practice my "Oh Oh Oh!" while bringing the Christmas gift? :)

Sure let's try that then ;)

**Comment by vttranlina** [comment]
## Updated tasks split

- [Implement DAO for MAIL Messages table](https://github.com/linagora/james-project/issues/4899)
- [PostgresUidProvider and PostgresModSeqProvider](https://github.com/linagora/james-project/issues/4931)
- Implement PostgresMessageMapper
- MailboxACL -> array in PostgresMailbox
- Guice bindings for PosgresMessageMapper
- Perf tests 

- [PostgresUsersDAO + PostgresUsersRepository](https://github.com/linagora/james-project/issues/4925)
- [PostgresDelegationStore](https://github.com/linagora/james-project/issues/4926)
- [PostgresDomainDAO + PostgresDomainList](https://github.com/linagora/james-project/issues/4924)
- [PostgresRecipientRewriteTableDAO + PostgresRecipientRewriteTable](https://github.com/linagora/james-project/issues/4927)
- [PostgresSieveQuotaDAO + PostgresSieveRepository](https://github.com/linagora/james-project/issues/4929)
- [PostgresSieveScriptDAO + PostgresSieveRepository](https://github.com/linagora/james-project/issues/4928)

**Comment by Arsnael** [comment]
I think we need to add PostgresAnnotationMapper too, necessary to bind PostgresMailboxMapper? => https://github.com/linagora/james-project/issues/4932
